Prepare SandboxExtensionHandle for generated serialization

achristensen07 opened the pull request:
#### aa4461b6312354f0bfe04f9acd39efac5b0e17fd
<pre>
Prepare SandboxExtensionHandle for generated serialization
<a href="https://bugs.webkit.org/show_bug.cgi?id=264302">https://bugs.webkit.org/show_bug.cgi?id=264302</a>
<a href="https://rdar.apple.com/118024534">rdar://118024534</a>

Reviewed by Chris Dumez.

A SandboxExtensionHandle is consumed when serializing, so it should be serialized as an r-value.
This means all structures that contain one must also be serialized as an r-value.
This PR just changes these structures to r-values with no changes in behavior except these 2:
1. Sending a WebPage::MouseEvent message was recently changed to send the sandbox extensions to
an iframe process if site isolation is enabled.  This restores behavior to what it was before that change.
2. In WebProcessPool::startMemorySampler we used to send a WebProcess::StartMemorySampler to each
web content process, but only the first would receive a sandbox extension handle.  This makes it so
that each process receives a sandbox extension handle.  This is a debug tool, so it shouldn&apos;t be impactful.

* Source/WebKit/GPUProcess/GPUProcessCreationParameters.serialization.in:
* Source/WebKit/GPUProcess/GPUProcessSessionParameters.serialization.in:
* Source/WebKit/NetworkProcess/NetworkProcessCreationParameters.serialization.in:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::encode):
(WebKit::NetworkSessionCreationParameters::encode const): Deleted.
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/Platform/IPC/FormDataReference.h:
(IPC::FormDataReference::encode const):
* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
(encode_type):
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder&lt;WebKit::RValueWithFunctionCalls&gt;::encode):
(IPC::ArgumentCoder&lt;WebKit::RValueWithFunctionCalls&gt;::decode):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
(WebKit::allSerializedTypes):
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
* Source/WebKit/Shared/Cocoa/SandboxExtensionCocoa.mm:
(WebKit::SandboxExtensionHandle::encode):
(WebKit::SandboxExtensionHandle::encode const): Deleted.
* Source/WebKit/Shared/Databases/IndexedDB/WebIDBResult.serialization.in:
* Source/WebKit/Shared/GoToBackForwardItemParameters.serialization.in:
* Source/WebKit/Shared/LoadParameters.cpp:
(WebKit::LoadParameters::encode):
(WebKit::LoadParameters::encode const): Deleted.
* Source/WebKit/Shared/LoadParameters.h:
* Source/WebKit/Shared/PolicyDecision.serialization.in:
* Source/WebKit/Shared/ResourceLoadStatisticsParameters.serialization.in:
* Source/WebKit/Shared/SandboxExtension.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/Shared/WebProcessDataStoreParameters.serialization.in:
* Source/WebKit/Shared/WebsiteDataStoreParameters.serialization.in:
* Source/WebKit/UIProcess/Cocoa/GPUProcessProxyCocoa.mm:
(WebKit::GPUProcessProxy::enablePowerLogging):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::grantAccessToAssetServices):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformInitializeWebProcess):
(WebKit::WebProcessPool::registerNotificationObservers):
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
(WebKit::WebProcessProxy::unblockAccessibilityServerIfNeeded):
* Source/WebKit/UIProcess/Downloads/DownloadProxy.cpp:
(WebKit::DownloadProxy::publishProgress):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::GPUProcessProxy):
(WebKit::GPUProcessProxy::updateSandboxAccess):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::sendCreationParametersToNewProcess):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::initializeWebPage):
(WebKit::ProvisionalPageProxy::goToBackForwardItem):
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::injectPageIntoNewProcess):
(WebKit::RemotePageProxy::sendMouseEvent):
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::finishGrantingRequest):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::loadRequestWithNavigationShared):
(WebKit::WebPageProxy::loadFile):
(WebKit::WebPageProxy::loadDataWithNavigationShared):
(WebKit::WebPageProxy::loadSimulatedRequest):
(WebKit::WebPageProxy::loadAlternateHTML):
(WebKit::WebPageProxy::reload):
(WebKit::WebPageProxy::sendMouseEvent):
(WebKit::WebPageProxy::receivedNavigationPolicyDecision):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processDidFinishLaunching):
(WebKit::WebProcessPool::startMemorySampler):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::shouldSendPendingMessage):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::resumeDownload):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/FileAPI/BlobRegistryProxy.cpp:
(WebKit::BlobRegistryProxy::registerInternalFileBlobURL):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::load):
* Source/WebKit/WebProcess/MediaStream/MediaDeviceSandboxExtensions.h:
(WebKit::MediaDeviceSandboxExtensions::ids):
(WebKit::MediaDeviceSandboxExtensions::handles):
(WebKit::MediaDeviceSandboxExtensions::machBootstrapHandle):
(WebKit::MediaDeviceSandboxExtensions::MediaDeviceSandboxExtensions): Deleted.
(WebKit::MediaDeviceSandboxExtensions::ids const): Deleted.
(WebKit::MediaDeviceSandboxExtensions::handles const): Deleted.
(WebKit::MediaDeviceSandboxExtensions::machBootstrapHandle const): Deleted.
* Source/WebKit/WebProcess/MediaStream/MediaDeviceSandboxExtensions.serialization.in:

Canonical link: <a href="https://commits.webkit.org/270323@main">https://commits.webkit.org/270323@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1561c309efc79f4f948007101ed8bad93ba670fb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25158 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3699 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26414 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27273 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23083 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5407 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1135 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23330 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25401 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2722 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21710 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27852 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/25320 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2408 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28771 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22954 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22997 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26594 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2363 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/654 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3709 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2800 "Built successfully") | | [⏳ 🛠 jsc-mips ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3209 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2695 "Built successfully") | | [⏳ 🧪 jsc-mips-tests ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
```diff
@@ -181,7 +181,7 @@ GPUProcessProxy::GPUProcessProxy()
+    send(Messages::GPUProcess::InitializeGPUProcess(WTFMove(parameters)), 0);
 
 #if PLATFORM(COCOA)
     m_hasSentGPUToolsSandboxExtensions = !parameters.gpuToolsExtensionHandles.isEmpty();
```
We're doing a use-after-move of parameters here.

cdumez commented on Source/WebKit/WebProcess/MediaStream/MediaDeviceSandboxExtensions.h:
```diff
@@ -47,9 +45,9 @@ class MediaDeviceSandboxExtensions {
-    const Vector<String>& ids() const { return m_ids; }
-    const Vector<SandboxExtension::Handle>& handles() const { return m_handles; }
-    const SandboxExtension::Handle& machBootstrapHandle() const { return m_machBootstrapHandle; }
+    Vector<String> ids() { return std::exchange(m_ids, { }); }
```
This is super confusing. This looks like a simple getter and yet it clears a data member. This needs to be renamed, takeIds() for e.g.

cdumez commented on Source/WebKit/WebProcess/MediaStream/MediaDeviceSandboxExtensions.h:
```diff
@@ -47,9 +45,9 @@ class MediaDeviceSandboxExtensions {
-    const Vector<SandboxExtension::Handle>& handles() const { return m_handles; }
-    const SandboxExtension::Handle& machBootstrapHandle() const { return m_machBootstrapHandle; }
+    Vector<String> ids() { return std::exchange(m_ids, { }); }
+    Vector<SandboxExtension::Handle> handles() { return std::exchange(m_handles, { }); }
```
ditto.

cdumez commented on Source/WebKit/WebProcess/MediaStream/MediaDeviceSandboxExtensions.h:
```diff
@@ -47,9 +45,9 @@ class MediaDeviceSandboxExtensions {
-    const SandboxExtension::Handle& machBootstrapHandle() const { return m_machBootstrapHandle; }
+    Vector<String> ids() { return std::exchange(m_ids, { }); }
+    Vector<SandboxExtension::Handle> handles() { return std::exchange(m_handles, { }); }
+    SandboxExtensionHandle machBootstrapHandle() { return std::exchange(m_machBootstrapHandle, { }); }
```
ditto.

cdumez reviewed (approved):
LGTM

achristensen07 committed:
Prepare SandboxExtensionHandle for generated serialization
https://bugs.webkit.org/show_bug.cgi?id=264302
rdar://118024534

Reviewed by Chris Dumez.

A SandboxExtensionHandle is consumed when serializing, so it should be serialized as an r-value.
This means all structures that contain one must also be serialized as an r-value.
This PR just changes these structures to r-values with no changes in behavior except these 2:
1. Sending a WebPage::MouseEvent message was recently changed to send the sandbox extensions to
an iframe process if site isolation is enabled.  This restores behavior to what it was before that change.
2. In WebProcessPool::startMemorySampler we used to send a WebProcess::StartMemorySampler to each
web content process, but only the first would receive a sandbox extension handle.  This makes it so
that each process receives a sandbox extension handle.  This is a debug tool, so it shouldn't be impactful.

* Source/WebKit/GPUProcess/GPUProcessCreationParameters.serialization.in:
* Source/WebKit/GPUProcess/GPUProcessSessionParameters.serialization.in:
* Source/WebKit/NetworkProcess/NetworkProcessCreationParameters.serialization.in:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::encode):
(WebKit::NetworkSessionCreationParameters::encode const): Deleted.
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/Platform/IPC/FormDataReference.h:
(IPC::FormDataReference::encode const):
* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
(encode_type):
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder<WebKit::RValueWithFunctionCalls>::encode):
(IPC::ArgumentCoder<WebKit::RValueWithFunctionCalls>::decode):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
(WebKit::allSerializedTypes):
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
* Source/WebKit/Shared/Cocoa/SandboxExtensionCocoa.mm:
(WebKit::SandboxExtensionHandle::encode):
(WebKit::SandboxExtensionHandle::encode const): Deleted.
* Source/WebKit/Shared/Databases/IndexedDB/WebIDBResult.serialization.in:
* Source/WebKit/Shared/GoToBackForwardItemParameters.serialization.in:
* Source/WebKit/Shared/LoadParameters.cpp:
(WebKit::LoadParameters::encode):
(WebKit::LoadParameters::encode const): Deleted.
* Source/WebKit/Shared/LoadParameters.h:
* Source/WebKit/Shared/PolicyDecision.serialization.in:
* Source/WebKit/Shared/ResourceLoadStatisticsParameters.serialization.in:
* Source/WebKit/Shared/SandboxExtension.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/Shared/WebProcessDataStoreParameters.serialization.in:
* Source/WebKit/Shared/WebsiteDataStoreParameters.serialization.in:
* Source/WebKit/UIProcess/Cocoa/GPUProcessProxyCocoa.mm:
(WebKit::GPUProcessProxy::enablePowerLogging):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::grantAccessToAssetServices):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformInitializeWebProcess):
(WebKit::WebProcessPool::registerNotificationObservers):
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
(WebKit::WebProcessProxy::unblockAccessibilityServerIfNeeded):
* Source/WebKit/UIProcess/Downloads/DownloadProxy.cpp:
(WebKit::DownloadProxy::publishProgress):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::GPUProcessProxy):
(WebKit::GPUProcessProxy::updateSandboxAccess):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::sendCreationParametersToNewProcess):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::initializeWebPage):
(WebKit::ProvisionalPageProxy::goToBackForwardItem):
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::injectPageIntoNewProcess):
(WebKit::RemotePageProxy::sendMouseEvent):
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::finishGrantingRequest):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::loadRequestWithNavigationShared):
(WebKit::WebPageProxy::loadFile):
(WebKit::WebPageProxy::loadDataWithNavigationShared):
(WebKit::WebPageProxy::loadSimulatedRequest):
(WebKit::WebPageProxy::loadAlternateHTML):
(WebKit::WebPageProxy::reload):
(WebKit::WebPageProxy::sendMouseEvent):
(WebKit::WebPageProxy::receivedNavigationPolicyDecision):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processDidFinishLaunching):
(WebKit::WebProcessPool::startMemorySampler):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::shouldSendPendingMessage):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::resumeDownload):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/FileAPI/BlobRegistryProxy.cpp:
(WebKit::BlobRegistryProxy::registerInternalFileBlobURL):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::load):
* Source/WebKit/WebProcess/MediaStream/MediaDeviceSandboxExtensions.h:
(WebKit::MediaDeviceSandboxExtensions::ids):
(WebKit::MediaDeviceSandboxExtensions::handles):
(WebKit::MediaDeviceSandboxExtensions::machBootstrapHandle):
(WebKit::MediaDeviceSandboxExtensions::MediaDeviceSandboxExtensions): Deleted.
(WebKit::MediaDeviceSandboxExtensions::ids const): Deleted.
(WebKit::MediaDeviceSandboxExtensions::handles const): Deleted.
(WebKit::MediaDeviceSandboxExtensions::machBootstrapHandle const): Deleted.
* Source/WebKit/WebProcess/MediaStream/MediaDeviceSandboxExtensions.serialization.in:

Canonical link: https://commits.webkit.org/270323@main
